### ISSUE TITLE
Question about enviromental variables

### ISSUE THREAD

**Issue by petobens** [winpython/winpython#48]
I've installed winpython today and I have a doubt about setting environmental variables. I understand that winpython aims to be a portable distribution but I wanted to be able to run python code from the `cmd.exe` and also use vim. So I've added the following directories to my `%PATH%`:

```
C:\winpython\python-3.4.2.amd64;C:\winpython\python-3.4.2.amd64\Scripts;C:\winpython\python-3.4.2.amd64\DLLs;C:\winpython\python-3.4.2.amd64\Lib
```

Now I can run scripts from the regular command prompt  but I'm not sure that what I've done is correct. Is there an automatic way of setting the path (I've registered the distribution but the path was not modified)? Should I add more directories besides the mentioned ones?

Strangely I can run vim from WInPython CommandPrompt but not from the regular cmd.exe (when using the official python distribution I could run vim from command prompt without problem). This leads me to believe that I might be missing something. Any help and guidance will be appreciated. Thanks in advance!

**Comment by stonebig** [comment]
Hi @petobens ,

You should have an example script in your
````
c:\winpython\scripts\env.bat
````

You can also launch 
````
c:\winpython\scripts\cmd.bat
````
to get a dos windows with the previous script executed.

**Comment by petobens** [comment]
Hi @stonebig , thanks for the quick reply. Let me first apologize for the fact that my following question is quite elementary but I can't get this to work. I want to permanently alter the `%PATH%`, adding the directories in `env.bat`. If I run `env.bat` I believe this will modifiy the `%PATH` temporarily because if after running the batch file I close the `cmd.exe` and re open it an run `python` I get the error saying it is not recognized as an internal command. 
So my question is: how do I permanently add to the `%PATH%` the same directories that `env.bat` adds temporarily?
Once again sorry for the inconvenience and thanks for the help.

**Comment by stonebig** [comment]
well,
looking at http://stackoverflow.com/questions/8358265/how-to-update-path-variable-permanently-from-cmd-windows

There is them manual change of the path method, described here
````
http://geekswithblogs.net/renso/archive/2009/10/21/how-to-set-the-windows-path-in-windows-7.aspx#540517````

Or there is this method, that I didn't try
````
rem (call winpython\script\cmd.bat, to launch the env.bat and have the dos command prompt)
rem then try this (from https://technet.microsoft.com/en-us/library/cc755104.aspx)
setx MYPATH %PATH%
````

Does anyone of this fit your problem ?

**Comment by petobens** [comment]
@stonebig  thanks for the references. I was able to set the user variable using the method described by the SO question (unfortunately I still have to set system variables manually because the `/M` flag suggested in SO doesn't seem to work at least in my case). Since you include the possibility to register the distribution and state that:

>these actions are exactly the same as those performed when installing Python with the official installer for Windows

and since the official installer gives the user the option (during installation) to automatically add the python executable to the `%PATH%` variable I was wondering if you could also add this option for winpython. 

One last question: I installed jedi with by doing `pip install jedi`, however `jedi` is not listed in the `uninstall package` list of the WinPython Control Panel. Is this correct?

One again thanks for the help!

**Comment by stonebig** [comment]
Yes, packages are listed only if you use  `WinPython Control Panel.exe`

You can do `pip uninstall jedi` then re-install it via `WinPython Control Panel.exe` if you want to see it.

**Comment by petobens** [comment]
Thanks for clarifying that point.  Regarding my other question,  will you consider adding an option to automatically add python to the path from within the winpython control panel (something the official python offers during installation ) ?  Or do you want me to close this issue?

**Comment by stonebig** [comment]
I didn't have time to think about that part of winpython yet, so I have no opinion. You can keep it open.

**Comment by petobens** [comment]
Ok. Thanks again for the help.

**Comment by ankostis** [comment]
+1 for adding WinPython in PATH, since without it, python-packages requiring DLLs fail.

**Comment by luigibrosse** [comment]
Hi, I would like to see something like this as well. It is quite tedious to manually insert / update python in the PATH each time that I move from one version of winpython to the next. Please consider it.

**Comment by yw5aj** [comment]
Being curious - is it that we are planning to do this in the future via changing the windows registry (http://stackoverflow.com/questions/2121795/programmatically-modifiy-environment-variables), or are we holding this off because chaning the registry will cause other bugs?

**Comment by petobens** [comment]
>PS: remember that we need to change HOME to settings dir as well... It's probably just me but I missed it earlier today.

@yw5aj if I change/create a `HOME` env variable and set it to the settings dir then when I fire up from the cmd prompt  a program that requires python, such as`vim`, such program doesn't know where to find its config file,`vimrc`, which lives in the  `%USERPROFILE%` directory (in my case, `C:\Users\Pedro`). 

The only way I found to get `gvim` (not `vim`)  to work is to create a `PYTHONHOME` env variable set to `C:\winpython\python-3.4.2.amd64`.  Your solution allows `vim` to work but I have no idea how to have it working and also loading the correct config file from my `%USERPROFILE%` directory. 

I don't know what the standard python installation does but when I chose to add python to the path then I could use both `vim` and `gvim` without problems (i.e loading the correcting config file).

Maybe this [link](http://www.mingw.org/wiki/HOWTO_Set_the_HOME_variable_for_cmd_exe) is relevant.

**Comment by yw5aj** [comment]
@petobens  Thanks! That's a good point. It almost look like that we want this:

When installing/configuring WinPython, we can select a irreversible option about whether to move the settings folder to the system %USERPROFILE% directory.

Not sure whether we'd like to do that, but that would solve the `gvim` issue (just an example) I guess?

Shawn

**Comment by petobens** [comment]
@yw5aj sorry for the late reply. 

I don't know if the solution is to move the settings  folder to the home directory (if I do that I believe that I can't open gvim or vim from the standard cmd prompt). 

What I (and others) want is just to have the option to add Winpython to the path env variable during installation (the way the regular python installer does). Hope @stonebig can implement this wanted feature.

**Comment by stonebig** [comment]
I still don't have a strong opinion on this and feel technically uncomfortable.

**Comment by petobens** [comment]
Okay.   No Problem!  Thanks for the quick reply.

**Comment by ankostis** [comment]
Dear @stonebig I receive at least once a month an "angry" request from people that i have recommended them using WinPython due to this issue.

The sequence of events is like this:

1. They want to run a python project of mine with native libs and DLLs (i.e. numpy & win32api).
2. I tell them "install WinPython and don't forget to manually update your PATH" .
3. They fail to do it :-) and i help them over the phone into achieving the above task.
4. Months later they *upgrade* their WinPython (i.e. to get more recent pandas).
5. THEY FORGET TO FIX THE PATH.
6. After days of trying to fix it, they come back shouting at me that my code is broken!

So, would it be possible to fix this once and for all?
It should have little to do with HOME and Settings.

**Comment by ankostis** [comment]
My suggestion of how this capability should work is this:

A) The *WinPython-installer* on start-up checks existing PATH for locating a (win)`python.exe`.

B.a) If `python.exe` exists in `PATH`, it provides this user-choice: 

  > Update python-X.Y.V --> Python-XX.YY.ZZ into your PATH environment variable ?

  and possibly add another one, whether to try and set this for all users).

B.b) If no `python.exe` exists in `PATH`, the choice should read: 

  > Add python-X.Y.V into your PATH environment variable?

C. Before exiting installer, if user had checked that option, it will:

  - i) gather all path-elements containing the identified (win)`python.exe`.
  - ii) Construct a ``setx PATH=...`` batch-script to *prepend* `PATH` with the new python (or use `win32api` registry-code) and 
  - iii) (depending on the 2nd choice) try running this with admin-rights or failback updating user-path env-var only.

I still have one issue unresolved, what to do if a `python.exe` from a different distribution is found (ie from *cygwin* or *anaconda*).

**Comment by stonebig** [comment]
Hi @ankostis,

What you suggest doesn't look a good idea to me.

Maybe Winpython is not currently the right solution for your users / your use case.

I didn't try it myself but maybe, maybe, pyzo would better fit your particular need ?

**Comment by stonebig** [comment]
I would think that the problem is nowodays solved per tweaking ./settings/winpython.ini
re-open if not.
